src/basler_parameters.cpp

```diff
@@ -0,0 +1,108 @@
+        if (!IsWritable(this_node))
+        {
+            ROS_ERROR_STREAM("Basler parameter '" << name << "' isn't writable or doesn't exist.");
+            ROS_ERROR_STREAM(name << " set to " << this_node->ToString());
+            return;
+        }
+        if (!IsAvailable(this_node->GetEntryByName(value.c_str())))
+        {
+            ROS_ERROR_STREAM("Value '" << value << "' isn't available for basler param '" << name << "'.");
+            return;
+        }
+        this_node->FromString(value.c_str());
+        ROS_DEBUG_STREAM(name << " set to " << this_node->ToString());
+    }
+    catch (const Pylon::GenericException& e)
+    {
+        ROS_ERROR_STREAM(e.GetDescription());
+    }
+}
+
+} // anonymous namespace
+
+void handle_basler_parameters(Pylon::CInstantCamera& camera)
+{
+    ros::NodeHandle private_handle("~");
+    XmlRpc::XmlRpcValue params;
+    private_handle.getParam("basler_params", params);
+    ROS_ASSERT_MSG(params.getType() == XmlRpc::XmlRpcValue::TypeArray, "Badly formed basler param yaml");
+    for (size_t index = 0; index < params.size(); ++index)
+    {
+        const auto& param = params[index];
+        ROS_ASSERT_MSG(param.getType() == XmlRpc::XmlRpcValue::TypeStruct, "Badly formed basler param yaml");
+        ROS_ASSERT_MSG(param.hasMember("name"), "Param needs name");
+        ROS_ASSERT_MSG(param.hasMember("type"), "Param needs type");
+        ROS_ASSERT_MSG(param.hasMember("value"), "Param needs value");
+        std::string name = param["name"];
+        std::string type = param["type"];
+        auto& value = param["value"];
+        if (type == "bool")
+        {
+            ROS_ASSERT_MSG(value.getType() == XmlRpc::XmlRpcValue::TypeBoolean,
+                           "Type of value for %s must be bool", name.c_str());
+            handle_basler_parameter<bool, GenApi::CBooleanPtr>(camera, name, value);
+        }
+        else if (type == "int")
+        {
+            ROS_ASSERT_MSG(value.getType() == XmlRpc::XmlRpcValue::TypeInt,
+                           "Type of value for %s must be int", name.c_str());
+            handle_basler_parameter<int, GenApi::CIntegerPtr>(camera, name, value);
+        }
+        else if (type == "float")
+        {
+            // xmlrpc uses doubles
+            ROS_ASSERT_MSG(value.getType() == XmlRpc::XmlRpcValue::TypeDouble,
+                           "Type of value for %s must be float", name.c_str());
+            handle_basler_parameter<double, GenApi::CFloatPtr>(camera, name, value);
+        }
+        else if (type == "enum")
+        {
+            ROS_ASSERT_MSG(value.getType() == XmlRpc::XmlRpcValue::TypeString,
+                           "Type of value for enum %s must be string", name.c_str());
+            handle_basler_enum_parameter(camera, name, value);
+        }
+        else
+        {
+            ROS_FATAL_STREAM("Unknown param type for parameter " << name << ": " << type);
```

Comment:
Why FATAL here instead of ERROR -- it looks like you use error above?  I feel like we normally use ERROR for this kind of thing.  I believe we only use FATAL right before killing the node (maybe that's the case here, but it'd be clearer if the FATAL statement is right before the node closes).

Reply:
This is from the original code - so far I've kept the behavior of the code in basler_parameters identical. I can change it to be error (and return false) though, and handle the exiting in the nodes.

Reply:
On 2nd thought - there's a lot of ROS_ASSERTs in this file (also all from pre-existing code), to catch unexpected yaml settings. This is also catching a bad yaml entry, so in that light, having a FATAL and crashing here seems more in line with the rest of the code - thoughts?

Reply:
+1 for crashing on bad inputs. I've forgotten - does ROS_FATAL also kill the node?

Reply:
I was looking and it doesn't seem like it does. I'm definitely not opposed to crashing the node on bad input -- just think it makes sense to do it in the same place as you announce the fatal error.  (If I misread documentation, though, that solves that problem).

Reply:
Just checked, and ROS_FATAL does kill the node (node no longer shows up in rosnode list). ROS_ASSERT however does not kill the node.